Flaky unit test: `TestCopyGraph_WithOptions/MountFrom_error`

The unit test `TestCopyGraph_WithOptions/MountFrom_error` in `copy_test.go` has been flaky in recent builds.
Here is an instance: https://github.com/oras-project/oras-go/actions/runs/13639868900/job/38127074239?pr=906

> === RUN   TestCopyGraph_WithOptions/MountFrom_error
>     copy_test.go:1810: count(Exists()) = 3, want 2
>     copy_test.go:1819: count(MountFrom()) = 2, want 1
> === RUN   TestCopyGraph_WithOptions/MountFrom_OnMounted_error
> --- FAIL: TestCopyGraph_WithOptions (0.01s)
>     --- PASS: TestCopyGraph_WithOptions/SkipNode (0.00s)
>     --- PASS: TestCopyGraph_WithOptions/MountFrom_mounted (0.00s)
>     --- PASS: TestCopyGraph_WithOptions/MountFrom_copied (0.00s)
>     --- PASS: TestCopyGraph_WithOptions/MountFrom_mounted_second_try (0.00s)
>     --- PASS: TestCopyGraph_WithOptions/MountFrom_copied_dst_not_a_Mounter (0.00s)
>     --- PASS: TestCopyGraph_WithOptions/MountFrom_empty_sourceRepositories (0.00s)
>     --- FAIL: TestCopyGraph_WithOptions/MountFrom_error (0.00s)
>     --- PASS: TestCopyGraph_WithOptions/MountFrom_OnMounted_error (0.00s)

https://github.com/oras-project/oras-go/blob/f3f718b3c423848c46b492669c7c433fb463b3fa/copy_test.go#L1792-L1821

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -71,19 +71,21 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	eg, egCtx := errgroup.WithContext(ctx)
+	// record the first occurred error only
 	var once sync.Once
 	var firstErr error
+	cancelWithErr := func(err error) {
+		once.Do(func() {
+			firstErr = err
+		})
+		cancel()
+	}
 
+	eg, egCtx := errgroup.WithContext(ctx)
 	for _, item := range items {
 		region := LimitRegion(egCtx, limiter)
 		if err := region.Start(); err != nil {
-			once.Do(func() {
-				// record the first error occurred
-				firstErr = err
-			})
-			// when an error occurs, cancel other tasks
-			cancel()
+			cancelWithErr(err)
 			// break loop instead of returning to allow already scheduled
 			// goroutines to finish their deferred region.End() calls
 			break
@@ -101,12 +103,7 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 				}
 
 				if err := fn(egCtx, lr, t); err != nil {
-					once.Do(func() {
-						// record the first error occurred
-						firstErr = err
-					})
-					// when an error occurs, cancel other tasks
-					cancel()
+					cancelWithErr(err)
 					return err
 				}
 				return nil
@@ -114,11 +111,9 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 		}(item, region))
 	}
 
-	if err := eg.Wait(); err != nil {
-		if firstErr != nil {
-			return firstErr
-		}
-		return err
+	egErr := eg.Wait()
+	if firstErr != nil {
+		return firstErr
 	}
-	return nil
+	return egErr
 }

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -63,18 +63,26 @@ func (lr *LimitedRegion) End() {
 	lr.ended = true
 }
 
+type goErr struct {
+	error
+}
+
+var emptyGoErr goErr
+
 // GoFunc represents a function that can be invoked by Go.
 type GoFunc[T any] func(ctx context.Context, region *LimitedRegion, t T) error
 
 // Go concurrently invokes fn on items.
 func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], items ...T) error {
 	eg, egCtx := errgroup.WithContext(ctx)
-	var egErr atomic.Value
+	var firstErr atomic.Value
+	firstErr.Store(emptyGoErr)
+
 	for _, item := range items {
 		region := LimitRegion(egCtx, limiter)
 		if err := region.Start(); err != nil {
-			if egErr, ok := egErr.Load().(error); ok && egErr != nil {
-				return egErr
+			if egErr, ok := firstErr.Load().(goErr); ok && egErr.error != nil {
+				return egErr.error
 			}
 			return err
 		}
@@ -83,7 +91,7 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 				defer region.End()
 				err := fn(egCtx, region, t)
 				if err != nil {
-					egErr.CompareAndSwap(nil, err)
+					firstErr.CompareAndSwap(nil, goErr{error: err})
 					return err
 				}
 				return nil

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -90,6 +90,8 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 			return func() error {
 				defer region.End()
 				err := fn(egCtx, region, t)
+				// TODO: remove egErr
+				// TODO: need to cancel egCtx before releasing (region.End)
 				if err != nil {
 					firstErr.CompareAndSwap(nil, goErr{error: err})
 					return err

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -82,30 +82,30 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 				// record the first error occurred
 				firstErr = err
 			})
-			// when error occurs, cancel other tasks
+			// when an error occurs, cancel other tasks
 			cancel()
 			// continue loop instead of returning to allow already scheduled
-			// goroutines to finish their deferred reg.End() calls
+			// goroutines to finish their deferred region.End() calls
 			continue
 		}
 
-		eg.Go(func(t T, reg *LimitedRegion) func() error {
+		eg.Go(func(t T, lr *LimitedRegion) func() error {
 			return func() error {
-				defer reg.End()
+				defer lr.End()
 
 				select {
 				case <-egCtx.Done():
-					// skip the task if the context is already canceled
+					// skip the task if the context is already cancelled
 					return nil
 				default:
 				}
 
-				if err := fn(egCtx, reg, t); err != nil {
+				if err := fn(egCtx, lr, t); err != nil {
 					once.Do(func() {
 						// record the first error occurred
 						firstErr = err
 					})
-					// when error occurs, cancel other tasks
+					// when an error occurs, cancel other tasks
 					cancel()
 					return err
 				}

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -17,6 +17,7 @@ package syncutil
 
 import (
 	"context"
+	"sync"
 
 	"golang.org/x/sync/errgroup"
 	"golang.org/x/sync/semaphore"
@@ -34,7 +35,6 @@ func LimitRegion(ctx context.Context, limiter *semaphore.Weighted) *LimitedRegio
 	if limiter == nil {
 		return nil
 	}
-	// initially marked as ended (i.e. not running)
 	return &LimitedRegion{
 		ctx:     ctx,
 		limiter: limiter,
@@ -67,26 +67,51 @@ func (lr *LimitedRegion) End() {
 type GoFunc[T any] func(ctx context.Context, region *LimitedRegion, t T) error
 
 // Go concurrently invokes fn on items.
+// It records the first “real” error (via sync.Once) and cancels the context.
+// Tasks that see cancellation before running f() return nil, so that Wait()
+// eventually returns your recorded error (if any).
 func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], items ...T) error {
-	// Create an explicit cancelable context so we can trigger cancellation immediately on error.
+	// Create an explicit cancelable context.
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	eg, egCtx := errgroup.WithContext(ctx)
+	var once sync.Once
+	var firstErr error
 
 	for _, item := range items {
 		region := LimitRegion(egCtx, limiter)
 		if err := region.Start(); err != nil {
-			return err
+			once.Do(func() {
+				firstErr = err
+			})
+			// Cancel other work and skip scheduling this task.
+			cancel()
+			// Instead of returning immediately, continue so that all previously
+			// scheduled goroutines can run their deferred reg.End() calls.
+			continue
 		}
 
-		// Capture the current item and region before launching the goroutine.
+		// Capture item and region so the closure gets its own copy.
 		eg.Go(func(t T, reg *LimitedRegion) func() error {
 			return func() error {
-				// When done (or upon error), release the semaphore permit.
+				// Always ensure the acquired permit is released.
 				defer reg.End()
+
+				// If the context is already canceled (by a previous error),
+				// skip executing fn() to avoid returning context.Canceled.
+				select {
+				case <-egCtx.Done():
+					return nil
+				default:
+				}
+
+				// Call the provided function.
 				if err := fn(egCtx, reg, t); err != nil {
-					// Cancel the context immediately so that other goroutines notice the failure.
+					once.Do(func() {
+						firstErr = err
+					})
+					// Cancel other goroutines.
 					cancel()
 					return err
 				}
@@ -94,5 +119,12 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 			}
 		}(item, region))
 	}
-	return eg.Wait()
+
+	if err := eg.Wait(); err != nil {
+		if firstErr != nil {
+			return firstErr
+		}
+		return err
+	}
+	return nil
 }

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -71,7 +71,7 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	// record the first occurred error only
+	// capture the first error occurred
 	var once sync.Once
 	var firstErr error
 	cancelWithErr := func(err error) {
@@ -86,7 +86,7 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 		region := LimitRegion(egCtx, limiter)
 		if err := region.Start(); err != nil {
 			cancelWithErr(err)
-			// break loop instead of returning to allow already scheduled
+			// break loop instead of returning to allow previously scheduled
 			// goroutines to finish their deferred region.End() calls
 			break
 		}
@@ -111,6 +111,7 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 		}(item, region))
 	}
 
+	// return the first meaningful error, avoiding "context cancelled"
 	egErr := eg.Wait()
 	if firstErr != nil {
 		return firstErr

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -17,7 +17,6 @@ package syncutil
 
 import (
 	"context"
-	"sync/atomic"
 
 	"golang.org/x/sync/errgroup"
 	"golang.org/x/sync/semaphore"
@@ -35,6 +34,7 @@ func LimitRegion(ctx context.Context, limiter *semaphore.Weighted) *LimitedRegio
 	if limiter == nil {
 		return nil
 	}
+	// initially marked as ended (i.e. not running)
 	return &LimitedRegion{
 		ctx:     ctx,
 		limiter: limiter,
@@ -68,29 +68,31 @@ type GoFunc[T any] func(ctx context.Context, region *LimitedRegion, t T) error
 
 // Go concurrently invokes fn on items.
 func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], items ...T) error {
+	// Create an explicit cancelable context so we can trigger cancellation immediately on error.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	eg, egCtx := errgroup.WithContext(ctx)
-	var egErr atomic.Value
+
 	for _, item := range items {
 		region := LimitRegion(egCtx, limiter)
 		if err := region.Start(); err != nil {
-			if egErr, ok := egErr.Load().(error); ok && egErr != nil {
-				return egErr
-			}
 			return err
 		}
-		eg.Go(func(t T) func() error {
+
+		// Capture the current item and region before launching the goroutine.
+		eg.Go(func(t T, reg *LimitedRegion) func() error {
 			return func() error {
-				defer region.End()
-				err := fn(egCtx, region, t)
-				// TODO: remove egErr
-				// TODO: need to cancel egCtx before releasing (region.End)
-				if err != nil {
-					egErr.CompareAndSwap(nil, err)
+				// When done (or upon error), release the semaphore permit.
+				defer reg.End()
+				if err := fn(egCtx, reg, t); err != nil {
+					// Cancel the context immediately so that other goroutines notice the failure.
+					cancel()
 					return err
 				}
 				return nil
 			}
-		}(item))
+		}(item, region))
 	}
 	return eg.Wait()
 }

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -17,7 +17,6 @@ package syncutil
 
 import (
 	"context"
-	"sync"
 
 	"golang.org/x/sync/errgroup"
 	"golang.org/x/sync/semaphore"
@@ -68,24 +67,14 @@ type GoFunc[T any] func(ctx context.Context, region *LimitedRegion, t T) error
 
 // Go concurrently invokes fn on items.
 func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], items ...T) error {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
-	// capture the first error occurred
-	var once sync.Once
-	var firstErr error
-	cancelWithErr := func(err error) {
-		once.Do(func() {
-			firstErr = err
-		})
-		cancel()
-	}
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(nil)
 
 	eg, egCtx := errgroup.WithContext(ctx)
 	for _, item := range items {
 		region := LimitRegion(egCtx, limiter)
 		if err := region.Start(); err != nil {
-			cancelWithErr(err)
+			cancel(err)
 			// break loop instead of returning to allow previously scheduled
 			// goroutines to finish their deferred region.End() calls
 			break
@@ -103,7 +92,7 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 				}
 
 				if err := fn(egCtx, lr, t); err != nil {
-					cancelWithErr(err)
+					cancel(err)
 					return err
 				}
 				return nil
@@ -111,10 +100,9 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 		}(item, region))
 	}
 
-	// return the first meaningful error, avoiding "context cancelled"
 	egErr := eg.Wait()
-	if firstErr != nil {
-		return firstErr
+	if causeErr := context.Cause(ctx); causeErr != nil {
+		return causeErr
 	}
 	return egErr
 }

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -63,26 +63,18 @@ func (lr *LimitedRegion) End() {
 	lr.ended = true
 }
 
-type goErr struct {
-	error
-}
-
-var emptyGoErr goErr
-
 // GoFunc represents a function that can be invoked by Go.
 type GoFunc[T any] func(ctx context.Context, region *LimitedRegion, t T) error
 
 // Go concurrently invokes fn on items.
 func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], items ...T) error {
 	eg, egCtx := errgroup.WithContext(ctx)
-	var firstErr atomic.Value
-	firstErr.Store(emptyGoErr)
-
+	var egErr atomic.Value
 	for _, item := range items {
 		region := LimitRegion(egCtx, limiter)
 		if err := region.Start(); err != nil {
-			if egErr, ok := firstErr.Load().(goErr); ok && egErr.error != nil {
-				return egErr.error
+			if egErr, ok := egErr.Load().(error); ok && egErr != nil {
+				return egErr
 			}
 			return err
 		}
@@ -93,7 +85,7 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 				// TODO: remove egErr
 				// TODO: need to cancel egCtx before releasing (region.End)
 				if err != nil {
-					firstErr.CompareAndSwap(nil, goErr{error: err})
+					egErr.CompareAndSwap(nil, err)
 					return err
 				}
 				return nil

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -68,7 +68,6 @@ type GoFunc[T any] func(ctx context.Context, region *LimitedRegion, t T) error
 
 // Go concurrently invokes fn on items.
 func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], items ...T) error {
-	// create an explicit cancelable context.
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
@@ -80,35 +79,33 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 		region := LimitRegion(egCtx, limiter)
 		if err := region.Start(); err != nil {
 			once.Do(func() {
+				// record the first error occurred
 				firstErr = err
 			})
-			// Cancel other work and skip scheduling this task.
+			// when error occurs, cancel other tasks
 			cancel()
-			// Instead of returning immediately, continue so that all previously
-			// scheduled goroutines can run their deferred reg.End() calls.
+			// continue loop instead of returning to allow already scheduled
+			// goroutines to finish their deferred reg.End() calls
 			continue
 		}
 
-		// Capture item and region so the closure gets its own copy.
 		eg.Go(func(t T, reg *LimitedRegion) func() error {
 			return func() error {
-				// Always ensure the acquired permit is released.
 				defer reg.End()
 
-				// If the context is already canceled (by a previous error),
-				// skip executing fn() to avoid returning context.Canceled.
 				select {
 				case <-egCtx.Done():
+					// skip the task if the context is already canceled
 					return nil
 				default:
 				}
 
-				// Call the provided function.
 				if err := fn(egCtx, reg, t); err != nil {
 					once.Do(func() {
+						// record the first error occurred
 						firstErr = err
 					})
-					// Cancel other goroutines.
+					// when error occurs, cancel other tasks
 					cancel()
 					return err
 				}

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -84,9 +84,9 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 			})
 			// when an error occurs, cancel other tasks
 			cancel()
-			// continue loop instead of returning to allow already scheduled
+			// break loop instead of returning to allow already scheduled
 			// goroutines to finish their deferred region.End() calls
-			continue
+			break
 		}
 
 		eg.Go(func(t T, lr *LimitedRegion) func() error {

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -100,9 +100,8 @@ func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], i
 		}(item, region))
 	}
 
-	egErr := eg.Wait()
-	if causeErr := context.Cause(ctx); causeErr != nil {
-		return causeErr
+	if err := eg.Wait(); err != nil {
+		cancel(err)
 	}
-	return egErr
+	return context.Cause(ctx)
 }

--- a/internal/syncutil/limit.go
+++ b/internal/syncutil/limit.go
@@ -67,11 +67,8 @@ func (lr *LimitedRegion) End() {
 type GoFunc[T any] func(ctx context.Context, region *LimitedRegion, t T) error
 
 // Go concurrently invokes fn on items.
-// It records the first “real” error (via sync.Once) and cancels the context.
-// Tasks that see cancellation before running f() return nil, so that Wait()
-// eventually returns your recorded error (if any).
 func Go[T any](ctx context.Context, limiter *semaphore.Weighted, fn GoFunc[T], items ...T) error {
-	// Create an explicit cancelable context.
+	// create an explicit cancelable context.
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 

--- a/internal/syncutil/limit_test.go
+++ b/internal/syncutil/limit_test.go
@@ -1,0 +1,82 @@
+package syncutil
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// TestLimitedRegionAllSuccess verifies that when no errors occur, all goroutines run to completion
+// and the semaphore is properly released.
+func TestLimitedRegionAllSuccess(t *testing.T) {
+	// Create a semaphore with capacity 2.
+	sem := semaphore.NewWeighted(2)
+	ctx := context.Background()
+	var counter int32
+
+	// Use numbers 1..5; our dummy function just sleeps a little and increments counter.
+	err := Go(ctx, sem, func(ctx context.Context, region *LimitedRegion, i int) error {
+		// Simulate work.
+		time.Sleep(10 * time.Millisecond)
+		atomic.AddInt32(&counter, 1)
+		return nil
+	}, 1, 2, 3, 4, 5)
+
+	if err != nil {
+		t.Fatalf("expected no error, got %v", err)
+	}
+
+	// After everything finishes, we expect counter to be 5.
+	if atomic.LoadInt32(&counter) != 5 {
+		t.Errorf("expected counter==5, got %d", counter)
+	}
+
+	// When all work is done the semaphore should have all permits available.
+	// Try acquiring the full weight—if the permits weren't all released,
+	// TryAcquire would return false.
+	if !sem.TryAcquire(2) {
+		t.Error("semaphore permits were not fully released at the end")
+	}
+	// Release the permits we just acquired.
+	sem.Release(2)
+}
+
+// TestLimitedRegionCancellation verifies that if an early error occurs,
+// (a) the overall Go call returns the error and (b) semaphore permits are released.
+func TestLimitedRegionCancellation(t *testing.T) {
+	sem := semaphore.NewWeighted(2)
+	ctx := context.Background()
+	var counter int32
+
+	// We choose a special item (e.g. value 42) to trigger error.
+	errTest := errors.New("intentional error")
+	err := Go(ctx, sem, func(ctx context.Context, region *LimitedRegion, i int) error {
+		// If we see the trigger value, return an error immediately.
+		if i == 42 {
+			return errTest
+		}
+		// Otherwise, simulate some work that takes time.
+		time.Sleep(50 * time.Millisecond)
+		atomic.AddInt32(&counter, 1)
+		return nil
+	}, 1, 42, 3, 4)
+
+	// In our design the first error should cancel work and be returned.
+	if err == nil {
+		t.Fatal("expected an error, but got nil")
+	}
+	if !errors.Is(err, errTest) {
+		t.Fatalf("expected error %v; got %v", errTest, err)
+	}
+
+	// Some of the non-error tasks may have been short-circuited.
+	// Regardless, once Go returns the semaphore should be fully released.
+	if !sem.TryAcquire(2) {
+		t.Error("semaphore permit was not released after error cancellation")
+	}
+	sem.Release(2)
+}

--- a/internal/syncutil/limit_test.go
+++ b/internal/syncutil/limit_test.go
@@ -46,26 +46,25 @@ func TestLimitedRegionAllSuccess(t *testing.T) {
 }
 
 // TestLimitedRegionCancellation verifies that if an early error occurs,
-// (a) the overall Go call returns the error and (b) semaphore permits are released.
+// the overall Go call returns the original error (instead of just context.Canceled)
+// and semaphore permits are released.
 func TestLimitedRegionCancellation(t *testing.T) {
 	sem := semaphore.NewWeighted(2)
 	ctx := context.Background()
 	var counter int32
 
-	// We choose a special item (e.g. value 42) to trigger error.
+	// We choose a special item (e.g. value 42) to trigger an error.
 	errTest := errors.New("intentional error")
 	err := Go(ctx, sem, func(ctx context.Context, region *LimitedRegion, i int) error {
-		// If we see the trigger value, return an error immediately.
 		if i == 42 {
 			return errTest
 		}
-		// Otherwise, simulate some work that takes time.
 		time.Sleep(50 * time.Millisecond)
 		atomic.AddInt32(&counter, 1)
 		return nil
 	}, 1, 42, 3, 4)
 
-	// In our design the first error should cancel work and be returned.
+	// We expect the first error to be errTest.
 	if err == nil {
 		t.Fatal("expected an error, but got nil")
 	}
@@ -73,8 +72,12 @@ func TestLimitedRegionCancellation(t *testing.T) {
 		t.Fatalf("expected error %v; got %v", errTest, err)
 	}
 
-	// Some of the non-error tasks may have been short-circuited.
-	// Regardless, once Go returns the semaphore should be fully released.
+	// After everything finishes, we expect counter to be smaller than 4.
+	if atomic.LoadInt32(&counter) >= 4 {
+		t.Errorf("expected counter < 4, got %d", counter)
+	}
+
+	// Ensure that the semaphore is fully released.
 	if !sem.TryAcquire(2) {
 		t.Error("semaphore permit was not released after error cancellation")
 	}

--- a/internal/syncutil/limit_test.go
+++ b/internal/syncutil/limit_test.go
@@ -70,7 +70,7 @@ func TestLimitedRegion_Cancellation(t *testing.T) {
 		return nil
 	}, 1, -1, 2, 0, -2)
 
-	// we expect the first error to be errTest.
+	// we expect the returned error to be errTest.
 	if !errors.Is(err, errTest) {
 		t.Fatalf("expected error %v; got %v", errTest, err)
 	}

--- a/internal/syncutil/limit_test.go
+++ b/internal/syncutil/limit_test.go
@@ -25,17 +25,13 @@ import (
 	"golang.org/x/sync/semaphore"
 )
 
-// TestLimitedRegionAllSuccess verifies that when no errors occur, all goroutines run to completion
-// and the semaphore is properly released.
-func TestLimitedRegionAllSuccess(t *testing.T) {
-	// Create a semaphore with capacity 2.
-	sem := semaphore.NewWeighted(2)
+func TestLimitedRegion_Success(t *testing.T) {
+	limiter := semaphore.NewWeighted(2)
 	ctx := context.Background()
 	var counter int32
 
-	// Use numbers 1..5; our dummy function just sleeps a little and increments counter.
-	err := Go(ctx, sem, func(ctx context.Context, region *LimitedRegion, i int) error {
-		// Simulate work.
+	err := Go(ctx, limiter, func(ctx context.Context, region *LimitedRegion, i int) error {
+		// just sleeps a little and increments counter to simulate task
 		time.Sleep(10 * time.Millisecond)
 		atomic.AddInt32(&counter, 1)
 		return nil
@@ -45,56 +41,48 @@ func TestLimitedRegionAllSuccess(t *testing.T) {
 		t.Fatalf("expected no error, got %v", err)
 	}
 
-	// After everything finishes, we expect counter to be 5.
-	if atomic.LoadInt32(&counter) != 5 {
-		t.Errorf("expected counter==5, got %d", counter)
+	// after everything finishes, we expect counter to be 5
+	if want := 5; atomic.LoadInt32(&counter) != int32(want) {
+		t.Errorf("expected counter == %v, got %v", want, counter)
 	}
 
-	// When all work is done the semaphore should have all permits available.
-	// Try acquiring the full weight—if the permits weren't all released,
-	// TryAcquire would return false.
-	if !sem.TryAcquire(2) {
+	// when all work is done the semaphore should have all permits available
+	if !limiter.TryAcquire(2) {
 		t.Error("semaphore permits were not fully released at the end")
 	}
-	// Release the permits we just acquired.
-	sem.Release(2)
+	limiter.Release(2)
 }
 
-// TestLimitedRegionCancellation verifies that if an early error occurs,
-// the overall Go call returns the original error (instead of just context.Canceled)
-// and semaphore permits are released.
-func TestLimitedRegionCancellation(t *testing.T) {
-	sem := semaphore.NewWeighted(2)
+func TestLimitedRegion_Cancellation(t *testing.T) {
+	limiter := semaphore.NewWeighted(2)
 	ctx := context.Background()
 	var counter int32
 
-	// We choose a special item (e.g. value 42) to trigger an error.
-	errTest := errors.New("intentional error")
-	err := Go(ctx, sem, func(ctx context.Context, region *LimitedRegion, i int) error {
-		if i == 42 {
+	errTest := errors.New("test error")
+	err := Go(ctx, limiter, func(ctx context.Context, region *LimitedRegion, i int) error {
+		if i < 0 {
+			// trigger an error on negative values
 			return errTest
 		}
+		// just sleeps a little and increments counter to simulate task
 		time.Sleep(50 * time.Millisecond)
 		atomic.AddInt32(&counter, 1)
 		return nil
-	}, 1, 42, 3, 4)
+	}, 1, -1, 2, 0, -2)
 
-	// We expect the first error to be errTest.
-	if err == nil {
-		t.Fatal("expected an error, but got nil")
-	}
+	// we expect the first error to be errTest.
 	if !errors.Is(err, errTest) {
 		t.Fatalf("expected error %v; got %v", errTest, err)
 	}
 
-	// After everything finishes, we expect counter to be smaller than 4.
-	if atomic.LoadInt32(&counter) >= 4 {
-		t.Errorf("expected counter < 4, got %d", counter)
+	// after everything finishes, we expect counter to be smaller than 5.
+	if max := 5; atomic.LoadInt32(&counter) >= int32(max) {
+		t.Errorf("expected counter < %v, got %v", max, counter)
 	}
 
-	// Ensure that the semaphore is fully released.
-	if !sem.TryAcquire(2) {
+	// when all work is done the semaphore should have all permits available
+	if !limiter.TryAcquire(2) {
 		t.Error("semaphore permit was not released after error cancellation")
 	}
-	sem.Release(2)
+	limiter.Release(2)
 }

--- a/internal/syncutil/limit_test.go
+++ b/internal/syncutil/limit_test.go
@@ -1,3 +1,18 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
 package syncutil
 
 import (

--- a/internal/syncutil/limitgroup.go
+++ b/internal/syncutil/limitgroup.go
@@ -21,7 +21,7 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
-// A LimitedGroup is a collection of goroutines working on subtasks that are part of
+// LimitedGroup is a collection of goroutines working on subtasks that are part of
 // the same overall task.
 type LimitedGroup struct {
 	grp *errgroup.Group
@@ -38,7 +38,9 @@ type LimitedGroup struct {
 // first.
 func LimitGroup(ctx context.Context, limit int) (*LimitedGroup, context.Context) {
 	grp, ctx := errgroup.WithContext(ctx)
-	grp.SetLimit(limit)
+	if limit > 0 {
+		grp.SetLimit(limit)
+	}
 	return &LimitedGroup{grp: grp, ctx: ctx}, ctx
 }
 
@@ -55,6 +57,7 @@ func (g *LimitedGroup) Go(f func() error) {
 		case <-g.ctx.Done():
 			return g.ctx.Err()
 		default:
+			// f() needs to check if the context is cancelled by itself?
 			return f()
 		}
 	})

--- a/internal/syncutil/limitgroup.go
+++ b/internal/syncutil/limitgroup.go
@@ -38,9 +38,7 @@ type LimitedGroup struct {
 // first.
 func LimitGroup(ctx context.Context, limit int) (*LimitedGroup, context.Context) {
 	grp, ctx := errgroup.WithContext(ctx)
-	if limit > 0 {
-		grp.SetLimit(limit)
-	}
+	grp.SetLimit(limit)
 	return &LimitedGroup{grp: grp, ctx: ctx}, ctx
 }
 
@@ -57,7 +55,6 @@ func (g *LimitedGroup) Go(f func() error) {
 		case <-g.ctx.Done():
 			return g.ctx.Err()
 		default:
-			// f() needs to check if the context is cancelled by itself?
 			return f()
 		}
 	})

--- a/internal/syncutil/limitgroup_test.go
+++ b/internal/syncutil/limitgroup_test.go
@@ -1,0 +1,128 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package syncutil
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+// TestLimitedGroupSuccess verifies that all functions complete without error.
+func TestLimitedGroupSuccess(t *testing.T) {
+	ctx := context.Background()
+	numTasks := 5
+
+	// Create a limited group with a concurrency limit of 2.
+	lg, _ := LimitGroup(ctx, 2)
+	var counter int32
+
+	for i := 0; i < numTasks; i++ {
+		// Capture i if needed.
+		lg.Go(func() error {
+			// Simulate some work.
+			time.Sleep(10 * time.Millisecond)
+			atomic.AddInt32(&counter, 1)
+			return nil
+		})
+	}
+
+	if err := lg.Wait(); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if got := atomic.LoadInt32(&counter); got != int32(numTasks) {
+		t.Errorf("expected counter %d, got %d", numTasks, got)
+	}
+}
+
+// TestLimitedGroupError confirms that if one task returns an error,
+// the act of canceling prevents further work and that Wait returns the original error.
+func TestLimitedGroupError(t *testing.T) {
+	ctx := context.Background()
+	lg, _ := LimitGroup(ctx, 2)
+	errTest := errors.New("intentional error")
+	var executed int32
+
+	// This task will eventually return an error.
+	lg.Go(func() error {
+		// Delay a bit so that other tasks are scheduled.
+		time.Sleep(20 * time.Millisecond)
+		atomic.AddInt32(&executed, 1)
+		return errTest
+	})
+
+	// This task simulates a slower, normal task.
+	lg.Go(func() error {
+		// Wait until cancellation is (hopefully) in effect.
+		time.Sleep(50 * time.Millisecond)
+		atomic.AddInt32(&executed, 1)
+		return nil
+	})
+
+	err := lg.Wait()
+	if !errors.Is(err, errTest) {
+		t.Fatalf("expected error %v, got %v", errTest, err)
+	}
+
+	// Since execution stops as soon as an error is encountered,
+	// it's possible that the second task never does its work.
+	if atomic.LoadInt32(&executed) < 1 {
+		t.Errorf("expected at least one task executed, got %d", executed)
+	}
+}
+
+// TestLimitedGroupLimit verifies that the concurrency limit is respected.
+func TestLimitedGroupLimit(t *testing.T) {
+	ctx := context.Background()
+	limit := 2
+	lg, _ := LimitGroup(ctx, limit)
+	var concurrent, maxConcurrent int32
+	numTasks := 10
+
+	for i := 0; i < numTasks; i++ {
+		lg.Go(func() error {
+			// Increment the count of concurrently active tasks.
+			cur := atomic.AddInt32(&concurrent, 1)
+			// Update the max concurrent tasks if needed.
+			for {
+				prevMax := atomic.LoadInt32(&maxConcurrent)
+				if cur > prevMax {
+					if atomic.CompareAndSwapInt32(&maxConcurrent, prevMax, cur) {
+						break
+					}
+				} else {
+					break
+				}
+			}
+
+			// Simulate a short task.
+			time.Sleep(20 * time.Millisecond)
+			atomic.AddInt32(&concurrent, -1)
+			return nil
+		})
+	}
+
+	if err := lg.Wait(); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if maxConcurrent > int32(limit) {
+		t.Errorf("expected max concurrent tasks <= %d, got %d", limit, maxConcurrent)
+	}
+}

--- a/internal/syncutil/limitgroup_test.go
+++ b/internal/syncutil/limitgroup_test.go
@@ -23,8 +23,7 @@ import (
 	"time"
 )
 
-// TestLimitedGroupSuccess verifies that all functions complete without error.
-func TestLimitedGroupSuccess(t *testing.T) {
+func TestLimitedGroup_Success(t *testing.T) {
 	ctx := context.Background()
 	numTasks := 5
 
@@ -32,10 +31,9 @@ func TestLimitedGroupSuccess(t *testing.T) {
 	lg, _ := LimitGroup(ctx, 2)
 	var counter int32
 
-	for i := 0; i < numTasks; i++ {
-		// Capture i if needed.
+	for range numTasks {
 		lg.Go(func() error {
-			// Simulate some work.
+			// simulate some work.
 			time.Sleep(10 * time.Millisecond)
 			atomic.AddInt32(&counter, 1)
 			return nil
@@ -51,25 +49,22 @@ func TestLimitedGroupSuccess(t *testing.T) {
 	}
 }
 
-// TestLimitedGroupError confirms that if one task returns an error,
-// the act of canceling prevents further work and that Wait returns the original error.
-func TestLimitedGroupError(t *testing.T) {
+func TestLimitedGroup_Error(t *testing.T) {
 	ctx := context.Background()
 	lg, _ := LimitGroup(ctx, 2)
-	errTest := errors.New("intentional error")
+	errTest := errors.New("test error")
 	var executed int32
 
-	// This task will eventually return an error.
 	lg.Go(func() error {
-		// Delay a bit so that other tasks are scheduled.
+		// delay a bit so that other tasks are scheduled.
 		time.Sleep(20 * time.Millisecond)
 		atomic.AddInt32(&executed, 1)
 		return errTest
 	})
 
-	// This task simulates a slower, normal task.
+	// simulates a slower, normal task.
 	lg.Go(func() error {
-		// Wait until cancellation is (hopefully) in effect.
+		// wait until cancellation is (hopefully) in effect.
 		time.Sleep(50 * time.Millisecond)
 		atomic.AddInt32(&executed, 1)
 		return nil
@@ -80,26 +75,23 @@ func TestLimitedGroupError(t *testing.T) {
 		t.Fatalf("expected error %v, got %v", errTest, err)
 	}
 
-	// Since execution stops as soon as an error is encountered,
-	// it's possible that the second task never does its work.
 	if atomic.LoadInt32(&executed) < 1 {
 		t.Errorf("expected at least one task executed, got %d", executed)
 	}
 }
 
-// TestLimitedGroupLimit verifies that the concurrency limit is respected.
-func TestLimitedGroupLimit(t *testing.T) {
+func TestLimitedGroup_Limit(t *testing.T) {
 	ctx := context.Background()
 	limit := 2
 	lg, _ := LimitGroup(ctx, limit)
 	var concurrent, maxConcurrent int32
 	numTasks := 10
 
-	for i := 0; i < numTasks; i++ {
+	for range numTasks {
 		lg.Go(func() error {
-			// Increment the count of concurrently active tasks.
+			// increment the count of concurrently active tasks.
 			cur := atomic.AddInt32(&concurrent, 1)
-			// Update the max concurrent tasks if needed.
+			// update the max concurrent tasks if needed.
 			for {
 				prevMax := atomic.LoadInt32(&maxConcurrent)
 				if cur > prevMax {
@@ -111,7 +103,7 @@ func TestLimitedGroupLimit(t *testing.T) {
 				}
 			}
 
-			// Simulate a short task.
+			// simulate a short task.
 			time.Sleep(20 * time.Millisecond)
 			atomic.AddInt32(&concurrent, -1)
 			return nil

